### PR TITLE
Borrow and modify SES delivery receipts implementation from GC Notify

### PR DESCRIPTION
**Pull request by jimmoffet**
Borrowed some delivery receipts handling infrastructure from GC Notify, which is partially described [here](https://github.com/cds-snc/notification-api/blob/main/CHANGELOG.md), under "Added a status callback for emails sent through SES"

Had to do a bit of additional integration, given that we branched from UK Notify several years later and GC Notify has had a few years of maintenance and development itself since the commits linked above.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -71,6 +71,7 @@ test: ## Run tests
 freeze-requirements: ## Pin all requirements including sub dependencies into requirements.txt
 	pip install --upgrade pip-tools
 	pip-compile requirements.in
+	pip3 install -r requirements.txt
```
Comment (jimmoffet) [inline Makefile:75]: Hard to imagine using this dev fixture without wanting to install the new dependency in the local dev environment...

File: app/celery/process_ses_receipts_tasks.py [hunk 2]
```diff
@@ -1,80 +1,176 @@
...
+    except decoder.JSONDecodeError:
+        current_app.logger.exception(f"Response headers: {request.headers}\nResponse data: {request.data}")
+        raise InvalidRequest("SES-SNS callback failed: invalid JSON given", 400)
+
+    current_app.logger.info(f"Message type: {message_type}\nResponse data: {message}")
+
+    try:
+        print("attempting to validate sns")
+        if valid_sns_message(message) == False:
+            current_app.logger.error(f"SES-SNS callback failed: validation failed! Response headers: {request.headers}\nResponse data: {request.data}\nError: Signature validation failed.")
```
Comment (jimmoffet) [inline app/celery/process_ses_receipts_tasks.py:73]: `valid_sns_message()` replaces the validatesns lib which seems to be no longer maintained at this point.

I ran into several issues with it and found [this thread](https://github.com/boto/boto3/issues/2508) which illuminated all of them.

IMO, the resulting function is fairly compact, easy to read, maintainable, and at least as defensive as the lib it replaces.

File: devcontainer-api/Dockerfile [hunk 3]
```diff
@@ -21,6 +21,7 @@ RUN apt-get update \
     openssh-client \
     procps \
     sudo \
+    swig \
     tldr \
```
Comment (jimmoffet) [inline devcontainer-api/Dockerfile:25]: required by M2Crypto, which provides cert validation

File: requirements.in [hunk 4]
```diff
@@ -16,10 +16,10 @@ itsdangerous==2.1.2
 jsonschema[format]==4.5.1
 marshmallow-sqlalchemy==0.28.1
 marshmallow==3.15.0
+M2Crypto==0.38.0
```
Comment (jimmoffet) [inline requirements.in:19]: Provides X509 standard cert validation for AWS .pem files (for validating SNS delivery receipt signatures). 20 years old and well-maintained.

File: app/celery/process_ses_receipts_tasks.py [hunk 5]
```diff
@@ -1,80 +1,166 @@
...
+        raise InvalidRequest("SES-SNS callback failed: invalid message type", 400)
+
+    try:
+        message = json.loads(request.data.decode('utf-8'))
+    except decoder.JSONDecodeError:
+        current_app.logger.exception(f"Response headers: {request.headers}\nResponse data: {request.data}")
+        raise InvalidRequest("SES-SNS callback failed: invalid JSON given", 400)
+
+    try:
+        validate_sns_message(message)
```
Comment (jimmoffet) [inline app/celery/process_ses_receipts_tasks.py:69]: validate_sns_message() replaces the validatesns lib, imported from GC Notify, which seems to be no longer maintained at this point.

I ran into several issues with it and found https://github.com/boto/boto3/issues/2508 which illuminated all of them.

I had to replace M2crypto in the above example with oscrypto because the linux base image (Ubuntu Bionic 18.04) used by CG does not have a backport for the current version of M2crypto. Oscrypto is widely-used and well-maintained.

File: requirements.in [hunk 6]
```diff
@@ -25,6 +25,7 @@ lxml==4.9.1
 defusedxml==0.7.1
 Werkzeug==2.1.1
 python-dotenv==0.20.0
+oscrypto==1.3.0
```
Comment (jimmoffet) [inline requirements.in:28]: Used to validate the cert that SNS sends with delivery receipt messages

File: app/celery/process_ses_receipts_tasks.py (hunk 5 above)
Comment (jimmoffet) [reply app/celery/process_ses_receipts_tasks.py:69]: FYI, if we _really_ needed to, we could validate sns certs using only openSSL

File: app/celery/validate_sns.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+    # Key signing cert url via Lambda and via webhook are slightly different
+    signing_cert_url = sns_payload.get('SigningCertUrl') if 'SigningCertUrl' in sns_payload else sns_payload.get('SigningCertURL')
+    if not isinstance(signing_cert_url, str):
+        raise ValidationError("Signing cert url must be a string")
+    cert_scheme, cert_netloc, *_ = urlparse(signing_cert_url)
+    if cert_scheme != 'https' or not re.match(_cert_url_re, cert_netloc):
+        raise ValidationError("Cert does not appear to be from AWS")
+    
+    certificate = _signing_cert_cache.get(signing_cert_url)
```
Comment (rahearn) [inline app/celery/validate_sns.py:95]: [issue] it doesn't look like there's anywhere that `_signing_cert_cache` has anything assigned to it

**Review by rahearn — COMMENTED** [review]
I think overall I'm having a little problem with the code organization. Seems to me like things are jammed into celery modules that should be in the web, and there was an incomplete refactoring of some callbacks functions.

File: app/celery/process_ses_receipts_tasks.py [hunk 8]
```diff
@@ -1,80 +1,166 @@
...
+        raise InvalidMessageTypeException(f'{message_type} is not a valid message type.')
+
+
+# 400 counts as a permanent failure so SNS will not retry.
+# 500 counts as a failed delivery attempt so SNS will retry.
+# See https://docs.aws.amazon.com/sns/latest/dg/DeliveryPolicies.html#DeliveryPolicies
+# This should not be here, it used to be in notifications/notifications_ses_callback. It then
+# got refactored into a task, which is fine, but it created a circular dependency. Will need
+# to investigate why GDS extracted this into a lambda
+@ses_callback_blueprint.route('/notifications/email/ses', methods=['POST'])
```
[the remark this answers is not in the document]
Comment (stvnrlly) [reply app/celery/process_ses_receipts_tasks.py]: That's a decorator, which is used to extend the function it's attached to in a predictable way. Some straightforward examples are [in the commands](https://github.com/GSA/notifications-api/blob/main/app/commands.py), where it sets up the `click` options to be passed into each command, or throughout the test suite, where it sets up the myriad `pytest` options for each test function.

Here, although I'm not fully back up to speed on [Flask blueprints](https://flask.palletsprojects.com/en/2.2.x/blueprints/), I believe that it's applying the blueprint route shortcuts to the function to create both an endpoint and (along with L29-33) attach error handling. It's a shortcut to a common pattern, basically. And yes, L26 creates the Blueprint object, which is then used here.

File: app/celery/process_ses_receipts_tasks.py [hunk 9]
```diff
@@ -1,80 +1,100 @@
...
+        if notification.status not in {NOTIFICATION_SENDING, NOTIFICATION_PENDING}:
             notifications_dao._duplicate_update_warning(
-                notification=notification,
-                status=notification_status
+                notification, 
+                notification_status
             )
             return
+
+        notifications_dao._update_notification_status(
```
Comment (rahearn) [inline app/celery/process_ses_receipts_tasks.py:77]: [python newbie question] does it mean anything when a method starts with a `_`? Either enforced by the language or by convention.

File: app/notifications/receive_notifications.py [hunk 10]
```diff
@@ -23,6 +24,64 @@
...
+    """
+    {
+        "originationNumber":"+14255550182",
+        "destinationNumber":"+12125550101",
+        "messageKeyword":"JOIN", # unique to our sending number
+        "messageBody":"EXAMPLE",
+        "inboundMessageId":"cae173d2-66b9-564c-8309-21f858e9fb84",
+        "previousPublishedMessageId":"wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
+    }
+    """
```
Comment (rahearn) [inline app/notifications/receive_notifications.py:38]: [question] is this an example of the expected payload? Can you label it as such if so?

**Review by rahearn — CHANGES_REQUESTED** [review]

**Comment by rahearn** [comment]
Meant to say this in my review comment, but I _really_ like the organization of this version. Much more clear to follow what's going on.

File: app/celery/process_ses_receipts_tasks.py [hunk 11]
```diff
@@ -1,80 +1,166 @@
...
+
+
+# 400 counts as a permanent failure so SNS will not retry.
+# 500 counts as a failed delivery attempt so SNS will retry.
+# See https://docs.aws.amazon.com/sns/latest/dg/DeliveryPolicies.html#DeliveryPolicies
+# This should not be here, it used to be in notifications/notifications_ses_callback. It then
+# got refactored into a task, which is fine, but it created a circular dependency. Will need
+# to investigate why GDS extracted this into a lambda
+@ses_callback_blueprint.route('/notifications/email/ses', methods=['POST'])
+def sns_callback_handler():
```
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/celery/process_ses_receipts_tasks.py]: refactored

File: app/celery/validate_sns.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+import oscrypto.asymmetric
+import oscrypto.errors
+import requests
+import six
+
+from app import redis_store
+from app.config import Config
+
+USE_CACHE = True
```
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/celery/validate_sns.py]: refactored

File: app/celery/validate_sns.py (hunk 7 above)
Comment (jimmoffet) [reply app/celery/validate_sns.py]: refactored

File: app/notifications/callbacks.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
+from app.celery.service_callback_tasks import send_delivery_status_to_service
+from app.config import QueueNames
+from app.dao.service_callback_api_dao import (
+    get_service_delivery_status_callback_api_for_service,
+)
+
+
+def check_and_queue_callback_task(notification):
```
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/notifications/callbacks.py]: refactored

File: app/notifications/callbacks.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    # queue callback task only if the service_callback_api exists
+    service_callback_api = get_service_delivery_status_callback_api_for_service(service_id=notification.service_id)
+    if service_callback_api:
+        notification_data = create_delivery_status_callback_data(notification, service_callback_api)
+
+        send_delivery_status_to_service.apply_async([str(notification.id), notification_data], queue=QueueNames.CALLBACKS)
+
+
+def create_delivery_status_callback_data(notification, service_callback_api):
+    from app import encryption
```
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/notifications/callbacks.py]: This is a bad code smell imported from GC Notify. It's been refactored out

File: app/celery/process_ses_receipts_tasks.py [hunk 15]
```diff
@@ -1,80 +1,100 @@
+import enum
+import traceback
 from datetime import datetime, timedelta
+from json import decoder
 
 import iso8601
+import requests
```
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/celery/process_ses_receipts_tasks.py]: done

File: app/celery/process_ses_receipts_tasks.py (hunk 9 above)
Comment (jimmoffet) [reply app/celery/process_ses_receipts_tasks.py:72]: No enforcement that I know of, but by convention these are often used to differentiate public from private/internal methods at a glance. Looks like this is an example where it was originally written as an internal method, but then started to get accessed elsewhere and was never renamed

File: app/notifications/sns_cert_validator.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+import oscrypto.asymmetric
+import oscrypto.errors
+import requests
+import six
+
+from app import redis_store
+from app.config import Config
+
+USE_CACHE = True
```
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/notifications/sns_cert_validator.py]: removed

File: app/notifications/sns_cert_validator.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+    # Key signing cert url via Lambda and via webhook are slightly different
+    signing_cert_url = sns_payload.get('SigningCertUrl') if 'SigningCertUrl' in sns_payload else sns_payload.get('SigningCertURL')
+    if not isinstance(signing_cert_url, str):
+        raise ValidationError("Signing cert url must be a string")
+    cert_scheme, cert_netloc, *_ = urlparse(signing_cert_url)
+    if cert_scheme != 'https' or not re.match(_cert_url_re, cert_netloc):
+        raise ValidationError("Cert does not appear to be from AWS")
+    
+    certificate = _signing_cert_cache.get(signing_cert_url)
```
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/notifications/sns_cert_validator.py:92]: In the first run, certificate will be empty, but since it is global to the function, it should remain available until it becomes inaccessible. Once it gets set in L97 via redis, it _should_ prevent redis from getting pinged again until a different url is requested, or until the current worker gets recycled.

File: app/celery/process_ses_receipts_tasks.py (hunk 5 above)
Comment (jimmoffet) [reply app/celery/process_ses_receipts_tasks.py]: function has been renamed validate_sns_cert and now lives in app.notifications.sns_cert_validator, but the above comments still apply

File: app/notifications/sns_cert_validator.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/notifications/sns_cert_validator.py:93]: Yep! Missed in the last refactor, added back in L34. Should now fill _signing_cert_cache on each call to redis

File: app/celery/process_ses_receipts_tasks.py [hunk 18]
```diff
@@ -84,5 +104,126 @@ def process_ses_results(self, response):
...
 
     except Exception as e:
-        current_app.logger.exception('Error processing SES results: {}'.format(type(e)))
+        current_app.logger.exception("Error processing SES results: {}".format(type(e)))
         self.retry(queue=QueueNames.RETRY)
+
+def determine_notification_bounce_type(ses_message):
+    notification_type = ses_message["notificationType"]
+    if notification_type in ["Delivery", "Complaint"]:
+        return notification_type
```
[the remark this answers is not in the document]
Comment (jimmoffet) [reply app/celery/process_ses_receipts_tasks.py:108]: Should be done, accepted the code above and added an extra line to raise a KeyError if we get an unexpected type

File: app/notifications/receive_notifications.py (hunk 10 above)
Comment (jimmoffet) [reply app/notifications/receive_notifications.py:39]: done

File: app/celery/process_ses_receipts_tasks.py (hunk 9 above)
Comment (rahearn) [reply app/celery/process_ses_receipts_tasks.py:72]: :eye-twitching:

**Review by rahearn — APPROVED** [review]
🎉
